### ISSUE TITLE
Module setup fails during creation of customer address attributes

### ISSUE THREAD

**Issue by robaimes** [shipperhq/module-shipper#121]
## Problem:
Attempting to install the Shipper HQ module produces an error and setup cannot complete.

## Reproduction:

- Install Magento (2.4.4) with preconditions above
- `composer require shipperhq/module-shipper`
- `bin/magento setup:upgrade`

## Additional Info
The exact error message spat out by the command line during `bin/magento setup:upgrade` is as below:

```
Unable to apply data patch ShipperHQ\Shipper\Setup\Patch\Data\InstallDestTypeAttributes for module ShipperHQ_Shipper. Original exception message: SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was: INSERT INTO `eav_attribute` (`entity_type_id`, `attribute_code`) VALUES (?, ?)
```

The specific query is
```sql
INSERT INTO `eav_attribute` (`entity_type_id`, `attribute_code`) VALUES ('2', 'destination_type')
```

The issue is present from lines 84 through 89 of `Setup\Patch\Data\InstallDestTypeAttributes`, and the same issue applies to lines 106 through 112.

Removing the `$attribute->save()` call allows setup to continue as normal, but without the `used_in_forms` data. Perhaps this would be better as a separate patch?

**Comment by ibraheemnabeelfauzi** [comment]
Hello @robaimes could you please clarify why there are two different database in the .env file?

**Comment by robaimes** [comment]
@ibraheemnabeelfauzi Same database but 2 different connections. I've later realised that this error occurs regardless. Using only a single connection still yields me this error.

**Comment by ibraheemnabeelfauzi** [comment]
@robaimes  thanks and sorry for delay in the response. With that being said, can you please check and see if that's not the case? https://stagebit.com/magento-2/magento-debugging/sqlstatehy000-general-error-1205-lock-wait-timeout-exceeded-try-restarting-transaction-magento/

**Comment by robaimes** [comment]
@ibraheemnabeelfauzi This is a way around the issue, yes. But the Patch script (in this case) is what is creating the lock. I also wouldn't be able to do this during a production deployment like I can a development environment.

**Comment by ibraheemnabeelfauzi** [comment]
Thanks @robaimes  let me check into this further.

**Comment by ibraheemnabeelfauzi** [comment]
@robaimes thank you for your patience. I checked with my folks. This is a deadlock, we'll look into changing our code but it may not be the case as we have 1000s of merchant running this without having such issues. We'll have a look and determine a way! 

Thank you for raising this!

**Comment by robaimes** [comment]
If it helps, the following patch allows setup to continue as normal:

<details>
<summary>Patch</summary>

```patch
diff --git a/src/Setup/Patch/Data/AddAttributesToForm.php b/src/Setup/Patch/Data/AddAttributesToForm.php
new file mode 100644
index 0000000..e7f4c1a
--- /dev/null
+++ b/src/Setup/Patch/Data/AddAttributesToForm.php
@@ -0,0 +1,99 @@
+<?php
+/*
+ * ShipperHQ
+ *
+ * @category ShipperHQ
+ * @package ShipperHQ_Shipper
+ * @copyright Copyright (c) 2022 Zowta LTD and Zowta LLC (http://www.ShipperHQ.com)
+ * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ * @author ShipperHQ Team sales@shipperhq.com
+ */
+declare(strict_types=1);
+
+namespace ShipperHQ\Shipper\Setup\Patch\Data;
+
+use Magento\Customer\Setup\CustomerSetup;
+use Magento\Customer\Setup\CustomerSetupFactory;
+use Magento\Framework\DB\Ddl\Table;
+use Magento\Framework\Setup\ModuleDataSetupInterface;
+use Magento\Framework\Setup\Patch\DataPatchInterface;
+
+class AddAttributesToForm implements DataPatchInterface
+{
+    /**
+     * @var ModuleDataSetupInterface $moduleDataSetup
+     */
+    private $moduleDataSetup;
+
+    /**
+     * Customer setup factory
+     * @var CustomerSetupFactory
+     */
+    private $customerSetupFactory;
+
+    /**
+     * @var string[]
+     */
+    private $attributes = [
+        'destination_type',
+        'validation_status',
+    ];
+
+    /**
+     * @param ModuleDataSetupInterface $moduleDataSetup
+     */
+    public function __construct(
+        ModuleDataSetupInterface $moduleDataSetup,
+        CustomerSetupFactory     $customerSetupFactory
+    ) {
+        $this->moduleDataSetup = $moduleDataSetup;
+        $this->customerSetupFactory = $customerSetupFactory;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public static function getDependencies()
+    {
+        return [];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getVersion()
+    {
+        return '1.0.5';
+    }
+
+    /**
+     * Do Upgrade
+     * @return void
+     */
+    public function apply()
+    {
+        /** @var CustomerSetup $customerSetup */
+        $customerSetup = $this->customerSetupFactory->create(['setup' => $this->moduleDataSetup]);
+
+        foreach ($this->attributes as $attributeCode) {
+            // add attribute to form
+            $attribute = $customerSetup->getEavConfig()->getAttribute(
+                'customer_address',
+                $attributeCode
+            );
+            $attribute->setData('used_in_forms', [
+                'adminhtml_customer_address',
+            ]);
+
+            $attribute->save();
+        }
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function getAliases()
+    {
+        return [];
+    }
+}
diff --git a/src/Setup/Patch/Data/InstallDestTypeAttributes.php b/src/Setup/Patch/Data/InstallDestTypeAttributes.php
index 56b1c79..cd6adde 100644
--- a/src/Setup/Patch/Data/InstallDestTypeAttributes.php
+++ b/src/Setup/Patch/Data/InstallDestTypeAttributes.php
@@ -82,11 +82,6 @@ class InstallDestTypeAttributes implements DataPatchInterface
                 'comment'  => 'ShipperHQ Address Type'
             ];
             $customerSetup->addAttribute('customer_address', 'destination_type', $destinationTypeAddressAttr);
-
-            // add attribute to form
-            $attribute = $customerSetup->getEavConfig()->getAttribute('customer_address', 'destination_type');
-            $attribute->setData('used_in_forms', ['adminhtml_customer_address']);
-            $attribute->save();
         }

         $existingvalidationStatusAttribute = $customerSetup->getAttribute('customer_address', 'validation_status');
@@ -104,12 +99,6 @@ class InstallDestTypeAttributes implements DataPatchInterface
                 'comment'  => 'ShipperHQ Address Validation Status'
             ];
             $customerSetup->addAttribute('customer_address', 'validation_status', $validationStatusAddressAttr);
-
-            // add attribute to form
-            /** @var  $attribute */
-            $attribute = $customerSetup->getEavConfig()->getAttribute('customer_address', 'validation_status');
-            $attribute->setData('used_in_forms', ['adminhtml_customer_address']);
-            $attribute->save();
         }
     }


```

</details>

**Comment by bawwwa** [comment]
Hi @shqsupport 

We faced the same issue with the data patch. 

Magento 2.4.4 (EE)
PHP 8.1

The patch @robaimes provided fixed the issue. 

Thanks!

**Comment by wsajosh** [comment]
Hi all,

Thanks for raising this issue and providing the patch @robaimes

I've just released 20.52.0 of module-shipper which includes the patch as well as a couple of other fixes. Please update and it should resolve this issue
